[ Tahoe ] TestWebKitAPI.WebKit.ClickAutoFillButton (api-test) is a constant timeout

pxlcoder opened the pull request:
#### 4baaf965a295b0a222424659183705eebaa88745
<pre>
[ Tahoe ] TestWebKitAPI.WebKit.ClickAutoFillButton (api-test) is a constant timeout
<a href="https://bugs.webkit.org/show_bug.cgi?id=301201">https://bugs.webkit.org/show_bug.cgi?id=301201</a>
<a href="https://rdar.apple.com/163129797">rdar://163129797</a>

Reviewed by Abrar Rahman Protyasha.

TestWebKitAPI.WebKit.ClickAutoFillButton uses a fixed location to click the
AutoFill button. This location is no longer accurate with the redesigned controls
in macOS Tahoe.

Fix by dynamically obtaining the location of the AutoFill button.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ClickAutoFillButton.mm:
(-[ClickAutoFillButton webProcessPlugInBrowserContextController:didClearWindowObjectForFrame:inScriptWorld:]):
(-[ClickAutoFillButton webProcessPlugIn:didCreateBrowserContextController:]):
(-[ClickAutoFillButton webProcessPlugInBrowserContextController:didFinishLoadForFrame:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
((WebKit, ClickAutoFillButton)):

Canonical link: <a href="https://commits.webkit.org/303219@main">https://commits.webkit.org/303219@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/784eb42e3ad766077b42770b97721eb2f800bcda

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/4175 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42684 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/139193 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0f9f551e-89bc-4a62-8475-043d02c8a381) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/133554 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/4102 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3938 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/100667 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83546 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/fb94befc-f4f2-412b-854e-65b43f023c3d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134630 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2972 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117957 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81434 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/aa4b6520-63de-4b5f-a2ab-4f8870b8f712) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/2858 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/694 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82415 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/111584 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36090 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141839 "Built successfully") | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3844 "Build is in progress. Recent messages:OS: Tahoe (26.0), Xcode: 26.0.1; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36660 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109030 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3925 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3405 "Found 1 new test failure: imported/w3c/web-platform-tests/webvtt/rendering/cues-with-video/processing-model/selectors/cue_function/not_root_selector.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/109184 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2937 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114237 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/57055 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20485 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3898 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32656 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3730 "Built successfully") | | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/3990 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3859 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/ClickAutoFillButton.mm:
```diff
@@ -35,35 +36,47 @@
 }
 
-@interface ClickAutoFillButton : NSObject <WKWebProcessPlugIn>
+@interface ClickAutoFillButton : NSObject <WKWebProcessPlugIn, WKWebProcessPlugInLoadDelegate>
```
optional: Maybe you could inherit from `WebProcessPlugInWithInternals`? Then you would get `-webProcessPlugInBrowserContextController:didClearWindowObjectForFrame:inScriptWorld:` for free, and you could change `didCreateBrowserContextController:` slightly to fit. Something like this?

```objc
  @interface ClickAutoFillButton : WebProcessPlugInWithInternals <WKWebProcessPlugInLoadDelegate>
  @end

  @implementation ClickAutoFillButton

  - (void)webProcessPlugIn:(WKWebProcessPlugInController *)plugInController didCreateBrowserContextController:(WKWebProcessPlugInBrowserContextController *)browserContextController
  {
      [super webProcessPlugIn:plugInController didCreateBrowserContextController:browserContextController];

      WKBundlePageUIClientV3 client;
      zeroBytes(client);
      client.base.version = 3;
      client.didClickAutoFillButton = didClickAutoFillButton;
      WKBundlePageSetUIClient([browserContextController _bundlePageRef], &client.base);
  }

  - (void)webProcessPlugInBrowserContextController:(WKWebProcessPlugInBrowserContextController *)browserContextController didFinishLoadForFrame:(WKWebProcessPlugInFrame *)frame
  {
      // ... rest of the implementation
  }

  @end
```

pxlcoder replied:
Sadly I don't think this is possible due to the way `WebProcessPlugin` is loaded.

```
Undefined symbols for architecture arm64e:
  "_OBJC_CLASS_$_WebProcessPlugInWithInternals", referenced from:
      _OBJC_CLASS_$_ClickAutoFillButton in libTestWebKitAPI.a[392](UnifiedSource5-nonARC.o)
  "_OBJC_METACLASS_$_WebProcessPlugInWithInternals", referenced from:
      _OBJC_METACLASS_$_ClickAutoFillButton in libTestWebKitAPI.a[392](UnifiedSource5-nonARC.o)
```

aprotyas replied:
😢

pxlcoder committed:
[ Tahoe ] TestWebKitAPI.WebKit.ClickAutoFillButton (api-test) is a constant timeout
https://bugs.webkit.org/show_bug.cgi?id=301201
rdar://163129797

Reviewed by Abrar Rahman Protyasha.

TestWebKitAPI.WebKit.ClickAutoFillButton uses a fixed location to click the
AutoFill button. This location is no longer accurate with the redesigned controls
in macOS Tahoe.

Fix by dynamically obtaining the location of the AutoFill button.

* Tools/TestWebKitAPI/Tests/WebKitCocoa/ClickAutoFillButton.mm:
(-[ClickAutoFillButton webProcessPlugInBrowserContextController:didClearWindowObjectForFrame:inScriptWorld:]):
(-[ClickAutoFillButton webProcessPlugIn:didCreateBrowserContextController:]):
(-[ClickAutoFillButton webProcessPlugInBrowserContextController:didFinishLoadForFrame:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/UIDelegate.mm:
((WebKit, ClickAutoFillButton)):

Canonical link: https://commits.webkit.org/303219@main
